libhb/rendersub.c

```diff
@@ -115,6 +128,115 @@ hb_filter_object_t hb_filter_render_sub =
     .close         = hb_rendersub_close,
 };
 
+static void hb_box_vec_resize(hb_box_vec_t *vec, int size)
+{
+    hb_box_t *boxes = realloc(vec->boxes, size * sizeof(hb_box_t));
+    if (boxes == NULL)
+    {
+        return; // Error. Should never happen.
+    }
+    vec->boxes = boxes;
+    vec->size = size;
+}
+
+static inline int hb_box_intersect(const hb_box_t *a, const hb_box_t *b, int offset)
+{
+    return ((FFMIN(a->x2, b->x2) + offset - FFMAX(a->x1, b->x1)) >= 0) &&
+           ((FFMIN(a->y2, b->y2) + offset - FFMAX(a->y1, b->y1)) >= 0);
+}
+
+static inline void hb_box_union(hb_box_t *a, const hb_box_t *b)
+{
+    if (a->x1 > b->x1) { a->x1 = b->x1; }
+    if (a->y1 > b->y1) { a->y1 = b->y1; }
+    if (a->x2 < b->x2) { a->x2 = b->x2; }
+    if (a->y2 < b->y2) { a->y2 = b->y2; }
+}
+
+static inline void hb_box_clear(hb_box_t *box)
+{
+    box->x1 = box->y1 = box->x2 = box->y2 = 0;
+}
+
+static void hb_box_vec_merge(hb_box_vec_t *vec)
+{
+    if (vec->count < 2)
+    {
+        return;
+    }
+
+    for (int i = 0; i < vec->count - 1; i++)
+    {
+        hb_box_t *a = &vec->boxes[i];
+        for (int j = i + 1; j < vec->count; j++)
+        {
+            hb_box_t *b = &vec->boxes[j];
+            if (hb_box_intersect(a, b, 8))
+            {
+                hb_box_union(a, b);
+                hb_box_clear(b);
+            }
+        }
+    }
+}
+
+static void hb_box_vec_compact(hb_box_vec_t *vec)
+{
+    int j = 0, i = 0;
+    while (i < vec->count)
+    {
+        hb_box_t a = vec->boxes[i];
+        if (a.x1 == 0 && a.y1 == 0 &&
+            a.x2 == 0 && a.y2 == 0)
```

Comment:
Checking for `x2 == 0 &&  y2 == 0` should be sufficient. Also you could move the `i++` out of the if-else. But this is nitpicking.

It's unclear to me if libass could ever return empty bitmaps. Maybe it can be safer to discard them in `hb_box_vec_append` by checking for `x1 == x2 || y1 == y2` in `if (vec->size < vec->count + 1)`?

Reply:
I don't think so, but adding an additional check won't hurt performance.